src/topology/inseparable.lean

```diff
@@ -178,12 +179,217 @@ def specialization_preorder : preorder X :=
 
 variable {X}
 
+local attribute [instance] specialization_preorder
+
 /-- A continuous function is monotone with respect to the specialization preorders on the domain and
 the codomain. -/
-lemma continuous.specialization_monotone (hf : continuous f) :
-  @monotone _ _ (specialization_preorder X) (specialization_preorder Y) f :=
+lemma continuous.specialization_monotone (hf : continuous f) : monotone f :=
 λ x y h, h.map hf
 
+lemma closure_singleton_eq_Iic (x : X) : closure {x} = Iic x :=
+ext $ λ _, specializes_iff_mem_closure.symm
+
+/-- A subset `S` of a topological space is stable under specialization
+if `y ∈ S → x ∈ S` for all `y ⤳ x`. -/
+def stable_under_specialization (s : set X) : Prop :=
+∀ ⦃x y⦄, x ⤳ y → x ∈ s → y ∈ s
+
+/-- A subset `S` of a topological space is stable under specialization
+if `x ∈ S → y ∈ S` for all `y ⤳ x`. -/
+def stable_under_generalization (s : set X) : Prop :=
+∀ ⦃x y⦄, y ⤳ x → x ∈ s → y ∈ s
+
+lemma stable_under_specialization_iff_is_lower_set {s : set X} :
+  stable_under_specialization s ↔ is_lower_set s :=
+iff.rfl
+
+lemma stable_under_generalization_iff_is_upper_set {s : set X} :
+  stable_under_generalization s ↔ is_upper_set s :=
+iff.rfl
+
+lemma is_closed.stable_under_specialization {s : set X} (hs : is_closed s) :
+  stable_under_specialization s :=
+λ x y e, e.mem_closed hs
+
+lemma is_open.stable_under_generalization {s : set X} (hs : is_open s) :
+  stable_under_generalization s :=
+λ x y e, e.mem_open hs
+
+lemma stable_under_generalization.compl {s : set X} (hs : stable_under_generalization s) :
+  stable_under_specialization sᶜ :=
+λ x y e, mt (hs e)
+
+lemma stable_under_specialization.compl {s : set X} (hs : stable_under_specialization s) :
+  stable_under_generalization sᶜ :=
+λ x y e, mt (hs e)
+
+lemma stable_under_specialization_compl_iff {s : set X} :
+  stable_under_specialization sᶜ ↔ stable_under_generalization s :=
+⟨λ h, compl_compl s ▸ h.compl, λ h, h.compl⟩
+
+lemma stable_under_generalization_compl_iff {s : set X} :
+  stable_under_generalization sᶜ ↔ stable_under_specialization s :=
+⟨λ h, compl_compl s ▸ h.compl, λ h, h.compl⟩
```

Comment:
There is [is_upper_set_compl](https://leanprover-community.github.io/mathlib_docs/order/upper_lower.html#is_upper_set_compl)